test/exercises/diffusion/heat_2d_dist_block.chpl

```diff
@@ -0,0 +1,68 @@
+/*
+  A distributed 2D finite-difference heat/diffusion equation solver
+
+  Computation is executed over a 2D distributed array.
+  The array distribution is managed by the `Block` distribution.
+  The `forall` loop manages task creation and synchronization
+  across and within locales.
+*/
+
+import BlockDist.Block,
+       Time.stopwatch;
+
+// create a stopwatch to time kernel execution
+var t = new stopwatch();
+config const writeTime = false;
+
+// compile with `-sRunCommDiag=true` to see comm diagnostics
+use CommDiagnostics;
+config param RunCommDiag = false;
+
+// declare configurable constants with default values
+config const nx = 256,      // number of grid points in x
+             ny = 256,      // number of grid points in y
+             nt = 50,       // number of time steps
+             alpha = 0.25,  // diffusion constant
+             solutionStd = 0.222751; // known solution for the default parameters
+
+// define a distributed 2D domain and subdomain to describe the grid and its interior
+const Indices = Block.createDomain({0..<nx, 0..<ny}),
+      IndicesInner = Indices[1..<nx-1, 1..<ny-1];
+
+// define a distributed 2D array over the above domain
+var u: [Indices] real;
+
+// set up initial conditions
+u = 1.0;
+u[nx/4..nx/2, ny/4..ny/2] = 2.0;
+
+// start comm diag
+if RunCommDiag then startCommDiagnostics();
+
+// create a temporary copy of 'u' to store the previous time step
+var un = u;
+
+// iterate for 'nt' time steps
+t.start();
+for 1..nt {
+  // swap arrays to prepare for next time step
+  u <=> un;
+
+  // compute the FD kernel in parallel
+  forall (i, j) in IndicesInner do
+    u[i, j] = un[i, j] + alpha *
+      (un[i-1, j] + un[i, j-1] + un[i+1, j] + un[i, j+1] - 4 * un[i, j]);
```

Comment:
I don't think I would do this for this PR, but just to note:  We could also consider a "local access" version of this version that made all the `u[i, j]`/`un[i.j]` accesses localAccesses.

(Or, really, we should improve the compiler so that it does this optimization automatically—(i,j) is an index of IndicesInner which is a subdomain of Indices, which was used to declare u and un, therefore we know they're aligned and that localAccess is legal, so it would rewrite it that way under the hood).